### PR TITLE
Added control over using mpi4py by setting environment variable PYOPTSPARSE_REQUIRE_MPI

### PR DESCRIPTION
**Pull request by hschilling**
## Purpose
This PR fixes issue #117.

The goal is to give the user some control over using mpi4py. The user can either say they don't want to run in parallel or explicitly say they must run in parallel. Otherwise, pyoptsparse currently tries to run in parallel as needed but if a suitable mpi4py is not available, there is a fallback to run in serial.

## Type of change
What types of change is it?
- New feature (non-breaking change which adds functionality)

## Testing
One test is to set `PYOPTSPARSE_REQUIRE_MPI` to '0'. Any tests that use ParOpt should fail since ParOpt needs to run in parallel and the environment variable is telling pyoptsparse to run in serial. 

## Checklist
_Put an `x` in the boxes that apply._

- [x] I have run unit and regression tests which pass locally with my changes
- [ ] I have added new tests that prove my fix is effective or that my feature works
- [x] I have added necessary documentation

### CODE REVIEW

File: pyoptsparse/pyParOpt/ParOpt.py [hunk 1]
```diff
@@ -2,11 +2,29 @@
...
+# If PYOPTSPARSE_REQUIRE_MPI is unset, attempt to import mpi4py, but continue on failure
+# with a notification.
+else:
+    try:
+        from paropt import ParOpt as _ParOpt
+        from mpi4py import MPI
+    except ImportError:
+        _ParOpt = None
+
+# try:
```
Comment (nwu63) [inline pyoptsparse/pyParOpt/ParOpt.py:23]: Please remove the comment block here.

Comment (hschilling) [reply pyoptsparse/pyParOpt/ParOpt.py:23]: Sorry about leaving the comment in. Removed. 

I will also add a test.

I can install black. Not an issue at all.

Comment (hschilling) [reply pyoptsparse/pyParOpt/ParOpt.py]: I pushed up a new version with black formatting and also a test

**Comment by nwu63** [comment]
I just took a look at the failed tests.
- For python 2, you could just skip the test since `importlib` does not provide a function called `reload`. As this feature is meant to work with OpenMDAO which only supports py3, this is perfectly fine IMO. Something like a try/except for `importlib.reload`, and put a SkipTest on the except portion.
- For python 3, I forgot that we had removed ParOpt from our test images temporarily. Please skip ParOpt tests for now and I will re-enable them once we start testing ParOpt.

**Comment by hschilling** [comment]
Neil, I could support Python 2. I didn't realize you did still support that:
https://docs.python.org/2/library/functions.html#reload

I am confused about the comment about ParOpt. I see you have tests that use paropt, for example:

https://github.com/mdolab/pyoptsparse/blob/master/test/test_hs015.py#L226

Won't that test fail?

**Comment by hschilling** [comment]
So I could do:

```if sys.version_info[0] == 2:
    reload_func = reload
else:
    reload_func = importlib.reload

class TestNoMPI(unittest.TestCase):

    # Next 3 tests check how the environment variable affects importing MPI
    def test_require_mpi(self):
        os.environ["PYOPTSPARSE_REQUIRE_MPI"] = "1"
        import pyoptsparse.pyOpt_MPI

        reload_func(pyoptsparse.pyOpt_MPI)
        self.assertTrue(inspect.ismodule(pyoptsparse.pyOpt_MPI.MPI))
```

**Comment by nwu63** [comment]
We will drop python 2 support some time this year, but seeing that it's fairly easy to support both, I would suggest we do that with the approach you posted.

We have `SkipTest` in place if we cannot import the optimizer. As ParOpt is not currently installed in our testing images, those tests will simply be skipped. See [here](https://github.com/mdolab/pyoptsparse/blob/master/test/test_hs015.py#L74)

**Comment by hschilling** [comment]
Ah, I see now. I should have looked in that method.

**Comment by hschilling** [comment]
Made the changes based on your comments. Hopefully that does the trick!

**Comment by nwu63** [comment]
`flake8` does more than formatting checks -- it performs syntax and "coding practices" checks. The two warnings we had were
- the ParOpt object imported was unused. This is fine since we are using the import just to determine whether we want to skip the test.
- for Python 3, the py2 `reload` function was undefined. This is also fine and something we have to live with if we are to support both py2 and py3.

I've pushed up a new commit which added inline ignore for these specific flake8 warnings. flake8 should pass now

**Pull request merged**